[Feature] Mitre Att&ck Intelligence + adapt Framework

Desvelao opened the pull request:
### Description
- Added new `Intelligence` tab to the `Mitre Att&ck` module with information about the Mitre resources: groups, mitigations, tactics, techniques using the new Wazuh API endpoints.
- Adapt Framework tab to the new Wazuh API endpoints.
 
### How to test
- Intelligence tab:
  - Each Mitre resource should be navigable when clicking on its label
  - Each Mitre resource should display a table with information and the table should:
     - Paginate
     - Sort
     - Open the details in a flyout with more information about the selected resource
     - Search bar:
        - Display suggestions and works the searching
        - Let searching with a term
  - Global search:
    - Let searching with a term and display the results by Mitre resource. It only displays the top 5 results, if there are more of them, a button with `See more results` label is displayed.
      - Click on the result button: open the resource details in a flyout
      - Click on the `See more results` button, navigate to the resource and add a filter (in the search bar) and display a table with the results
- Intelligence detail flyout
  - Display basic information about the resource
  - Display another information in tables.
  - Click on another resource should refresh the details flyout with the information of the selected resource
  
- Framework
  - The Tactics and Techniques panels should display.
  - Open a technique should open a flyout:
    - Click on a technique should navigate to the Intelligence section and the selected resource as a filter in the resource search bar
    - Click on a rule should navigate the Rules with the selected rule
    
Limitations:
- The resource tables can't be sorted by `ID` due to a Wazuh API limitation https://github.com/wazuh/wazuh/issues/8879
Closes #3202

CPAlejandro commented:
When selecting a link from the description of any group/technique/tactics ... first a flyout opens and then it redirects you to the official miter page, perhaps it could be controlled so that in these cases, the flyout does not appear

From what I see, in the Intelligence tab, the agent selector is not used at all, perhaps it could be hidden. And when an agent is pinned the same.

Desvelao commented:
### Changes
- Modified how to open the resource details flyout
> Open details clicking in `ID` or `Name` resource properties instead of the entire row. This lets to use the links in the `description` column generated by the `markdown`
- Refactored some properties components
- Removed not used code
- Fix changelog

frankeros commented on public/components/common/hooks/use_async_action.ts:
```diff
@@ -0,0 +1,32 @@
+      setRunning(true);
+      const data = await action(...params);
+      setData(data);
+    }catch(error){
```
use prettier

frankeros replied:
desired unit test

Desvelao replied:
Added unit test

frankeros commented on public/components/common/panels/panel_split.tsx:
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { EuiPanel, EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
+
+export const PanelSplit = ({side, sideColor='#80808014', content, panelProps = {}, sideProps = {}, contentProps = {}}) => {
```
It would be nice the unit test for this component

frankeros commented on public/components/common/util/markdown.tsx:
```diff
@@ -0,0 +1,33 @@
+  className?: string
+};
+
+export const Markdown = ({markdown, className = ''}: MarkdownProps) => (
```
desirable unit test

Desvelao replied:
Added unit test

frankeros commented on public/components/common/util/markdown.tsx:
```diff
@@ -0,0 +1,33 @@
+export const Markdown = ({markdown, className = ''}: MarkdownProps) => (
+  <div
+    className={classnames('wz-markdown-margin', 'wz-markdown-wapper', className)}
+    dangerouslySetInnerHTML={{__html: md.render(markdown)}}>
```
we need to avoid using `dangerouslySetInnerHTML`, but if it is totally necessary we should use a sanitizer to avoid XSS
i.e. [DOMPurify](https://github.com/cure53/DOMPurify)

```js
import { sanitize } from 'dompurify';
...
 <div
    className={classnames('wz-markdown-margin', 'wz-markdown-wapper', className)}
    dangerouslySetInnerHTML={{__html: sanitize(md.render(markdown))}}>
  </div>
```

frankeros replied:
maybe is a good idea to leave a justification like Kibana does
![image](https://user-images.githubusercontent.com/26828184/122060167-b8f03980-cdc3-11eb-9211-9af41168335e.png)

Desvelao replied:
The content of the `markdown` comes from the Wazuh API response. Do we really need to add that dependency to the app to this use case?

Desvelao replied:
Added comment with the justification to using `dangerouslySetInnerHTML`.

frankeros commented on public/components/overview/mitre/components/techniques/components/flyout-technique/flyout-technique.tsx:
```diff
@@ -204,60 +187,45 @@ export class FlyoutTechnique extends Component {
+                <EuiToolTip
+                  position="top"
+                  content={
+                    "Open " + tactic.name + " details in a Intelligence section"
```
spell
`details in the Intelligence section`

Desvelao replied:
Fixed and replaced by template string

frankeros commented on public/components/overview/mitre/components/techniques/components/flyout-technique/flyout-technique.tsx:
```diff
@@ -320,7 +281,7 @@ export class FlyoutTechnique extends Component {
           <EuiFlexGroup className="flyout-row">
             <EuiFlexItem>
-              <Discover kbnSearchBar shareFilterManager={this.filterManager} initialColumns={["icon", "timestamp", 'rule.mitre.id', 'rule.mitre.tactic', 'rule.level', 'rule.id', 'rule.description']} implicitFilters={implicitFilters} initialFilters={[]} updateTotalHits={(total) => this.updateTotalHits(total)}/>
+              <Discover kbnSearchBar shareFilterManager={this.filterManager} initialColumns={["icon", "timestamp", 'rule.mitre.id', 'rule.mitre.tactic', 'rule.level', 'rule.id', 'rule.description']} implicitFilters={implicitFilters} initialFilters={[]} updateTotalHits={(total) => this.updateTotalHits(total)} openIntelligence={(e,redirectTo,itemId) => this.props.openIntelligence(e,redirectTo,itemI ...
```
prettier

Desvelao replied:
Fixed

frankeros commented on public/components/overview/mitre/components/techniques/techniques.tsx:
```diff
@@ -177,29 +189,73 @@ export const Techniques = withWindowSize(class Techniques extends Component {
+    mitreTechniques.push(...output.data.data.affected_items);
+    if (totalItems && output.data && output.data.data && totalItems > 500) {
+      params.offset = 0;
+      while (mitreTechniques.length < totalItems && params.offset < totalItems) {
+        params.offset += params.limit;
+        const tmpData = await this.getMitreTechniques(params)
+        mitreTechniques.push(...tmpData.data.data.affected_items);
+      }
```
if this is sync use `for` instead of `while` 
if this is async use `await Promise.all` 
using `while` with `await` inside may have bad behaviors

Desvelao replied:
Replaced to use with `Promise.all` and do the requests in parallel

frankeros commented on public/components/overview/mitre_attack_intelligence/module_mitre_attack_intelligence_all_resources.tsx:
```diff
@@ -0,0 +1,48 @@
+/*
```
I think the name of the files are too long
`mitre_attack_intelligence` is repeated from the parent folder maybe is not necessary for the name of the file

Desvelao replied:
Reduced the length of file names

frankeros commented on public/components/overview/mitre_attack_intelligence/module_mitre_attack_intelligence_all_resources.tsx:
```diff
@@ -0,0 +1,48 @@
+  EuiTitle,
+  EuiSpacer
+} from '@elastic/eui';
+import { ModuleMitreAttackIntelligenceAllResourcesSearchResults } from './module_mitre_attack_intelligence_all_resources_search_results';
```
same here too long

Desvelao replied:
Reduced the length of file names

frankeros commented on public/components/overview/mitre_attack_intelligence/module_mitre_intelligence.test.tsx:
```diff
@@ -0,0 +1,49 @@
+    expect(component).toMatchSnapshot();
+  });
+
+  // test('should render a Health check screen with error', () => {
+  //   const component = mount(<HealthCheckTest />);
+
+  //   component.find('CheckResult').at(1).invoke('handleErrors')('setup',['Test error']); // invoke is wrapped with act to await for setState
+
+  //   const callOutError = component.find('EuiCallOut');
+  //   expect(callOutError.text()).toBe('Test error');
+  // });
```
remove

Desvelao replied:
Removed

frankeros reviewed (changes requested):
there are several new components without unit test 
also, there are names of the files and components too long with the repeated name of theirs parent folder please keep it easy to read, I mean it must be between a descriptive but assertive name

CPAlejandro commented:
I have found this behavior, if we do a search and we see more results and then we change the table, we lose the filter. The same happens if we click on the table we were seeing again.

https://user-images.githubusercontent.com/17100196/122065530-4fd3ea80-cdf2-11eb-9bad-a70d4b227f21.mp4

Desvelao commented:
> When selecting a link from the description of any group/technique/tactics ... first a flyout opens and then it redirects you to the official Mitre page, perhaps it could be controlled so that in these cases, the flyout does not appear

Fixed. I changed how to open the resource details flyout to let clicking on the possible links that could contain the `description` column

> From what I see, in the Intelligence tab, the agent selector is not used at all, perhaps it could be hidden. And when an agent is pinned the same.

At the moment, we aren't modifing this. But make sense to do it.

CPAlejandro commented on public/components/overview/mitre/components/techniques/techniques.tsx:
```diff
@@ -177,29 +189,73 @@ export const Techniques = withWindowSize(class Techniques extends Component {
+  }
+
+   buildObjTechniques(techniques){
+    const techniquesObj = []
```
semicolon

Desvelao replied:
Added the semicolon

CPAlejandro commented on public/components/overview/mitre/components/techniques/techniques.tsx:
```diff
@@ -177,29 +189,73 @@ export const Techniques = withWindowSize(class Techniques extends Component {
+        mitreTechniqueID ? techniquesObj.push({ id : mitreTechniqueID, name: mitreTechniqueName}) : '';
+      }
+    });
+    return techniquesObj
```
semicolon

Desvelao replied:
Added the semicolon

Desvelao commented:
@frankeros , do we really want to simplify the name of components? I think they are descriptive, and I am not sure if we simplify them, it could collision with other possible future components or be not very descriptive, doing the searching in code of them harder. If I get an example of a name, I will rename them.

frankeros reviewed (approved):
LGTM! Great work!!

pablomarga commented:
When entering in Framework tab an error occurs (Sample data used).
![Screenshot_20210617_103044](https://user-images.githubusercontent.com/46006326/122361033-2d0f1680-cf57-11eb-9790-16ae1f09dd5f.png)

pablomarga reviewed (approved):
LGTM!
